hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngradeFixtures.java

```diff
@@ -0,0 +1,633 @@
+  
+  @TempDir
+  java.nio.file.Path tempDir;
+  
+  private HoodieTableMetaClient metaClient;
+
+  @BeforeEach
+  public void setUp() throws IOException {
+    initSparkContexts();
+  }
+
+  @AfterEach
+  public void tearDown() throws IOException {
+    cleanupSparkContexts();
+  }
+
+  @ParameterizedTest
+  @MethodSource("fixtureVersions")
+  public void testRoundTripUpgradeDowngrade(HoodieTableVersion originalVersion) throws Exception {
+    LOG.info("Testing round-trip upgrade/downgrade for version {}", originalVersion);
+    
+    // Load fixture table
+    HoodieTableMetaClient originalMetaClient = loadFixtureTable(originalVersion);
+    assertEquals(originalVersion, originalMetaClient.getTableConfig().getTableVersion(),
+        "Fixture table should be at expected version");
+    
+    // Calculate target version (next version up)
+    HoodieTableVersion targetVersion = getNextVersion(originalVersion);
+    if (targetVersion == null) {
+      LOG.info("Skipping upgrade test for version {} (no higher version available)", originalVersion);
+      return;
+    }
+    
+    // Create write config for upgrade operations
+    HoodieWriteConfig config = createWriteConfig(originalMetaClient, true);
+    
+    // Step 1: Upgrade to next version
+    LOG.info("Step 1: Upgrading from {} to {}", originalVersion, targetVersion);
+    new UpgradeDowngrade(originalMetaClient, config, context, SparkUpgradeDowngradeHelper.getInstance())
+        .run(targetVersion, null);
+    
+    // Create fresh meta client to read updated table configuration after upgrade
+    HoodieTableMetaClient upgradedMetaClient = HoodieTableMetaClient.builder()
+        .setConf(storageConf.newInstance())
+        .setBasePath(originalMetaClient.getBasePath())
+        .build();
+    assertTableVersionOnDataAndMetadataTable(upgradedMetaClient, targetVersion);
+    validateVersionSpecificProperties(upgradedMetaClient, originalVersion, targetVersion);
+    performDataValidationOnTable(upgradedMetaClient, "after upgrade");
+
+    // Step 2: Downgrade back to original version
+    LOG.info("Step 2: Downgrading from {} back to {}", targetVersion, originalVersion);
+    new UpgradeDowngrade(upgradedMetaClient, config, context, SparkUpgradeDowngradeHelper.getInstance())
+        .run(originalVersion, null);
+    
+    // Create fresh meta client to read updated table configuration after downgrade
+    HoodieTableMetaClient finalMetaClient = HoodieTableMetaClient.builder()
+        .setConf(storageConf.newInstance())
+        .setBasePath(upgradedMetaClient.getBasePath())
+        .build();
+    assertTableVersionOnDataAndMetadataTable(finalMetaClient, originalVersion);
+    validateVersionSpecificProperties(finalMetaClient, targetVersion, originalVersion);
+    performDataValidationOnTable(finalMetaClient, "after round-trip");
+
+    LOG.info("Successfully completed round-trip test for version {}", originalVersion);
```

Comment:
Should the downgrade tests be separated into a new test?  After upgrade, in certain cases the rollback is done so the upgraded table does not have inflight commit any more, thus the downgrade testing here does not cover the case with inflight commit and rollback.

Reply:
I think this is a good point let me look into separating these out so its clearer.